tests/tape/tapes/test_qnode.py

```diff
@@ -915,3 +915,21 @@ def circuit(x):
         # test differentiability. The circuit will assume an RZ gate
         grad = qml.grad(circuit)(-0.5)
         np.testing.assert_allclose(grad, 0, atol=tol, rtol=0)
+
+
+class TestShots:
+    """Unittests for specifying shots per call."""
+
+    def test_specify_shots(self):
+        """Tests that shots can be set per call."""
+        dev = qml.device('default.qubit', wires=1, shots=10)
+
+        @qml.qnode(dev)
+        def circuit(a):
+            qml.RX(a, wires=0)
+            return qml.sample(qml.PauliZ(wires=0))
+
+        assert len(circuit(0.8)) == 10
+        assert len(circuit(0.8, shots=2)) == 2
+        assert len(circuit(0.8, shots=3178)) == 3178
+        assert len(circuit(0.8)) == 10

```

Comment:
Are there other tests we might need?

Reply:
How about a test that ensure that the device remains immutable in the case of device swapping?

Reply:
Like that the device shots do not change? But wouldn't that be a test for the device swapping part of the code?

Reply:
I suppose so, yeah. But it's a good integration test, to make sure we aren't mutating the device without realising? The test could be parametrized over `parameter-shift` (no device swapping) and `backprop` (device swapping).